### PR TITLE
[FLINK-2458][FLINK-2449]Access distributed cache entries for CollectionExecution and in Iterative tasks.

### PR DESCRIPTION
**Pull request by sachingoel0101**
1. This PR adds support for accessing distributed cache entries when running iterations.
2. Since there are several tests which execute on both Cluster and Collection modes, it seems logical to not fail a test on either if it passes on both. Distributed Cache files create one such case. There is nothing actually wrong with trying to access a distributed cache entry when running in collection environment. It just doesn't really make sense to do so.
This takes care of that too.

### CODE REVIEW

File: flink-core/src/main/java/org/apache/flink/api/common/functions/util/AbstractRuntimeUDFContext.java [hunk 1]
```diff
@@ -79,7 +68,7 @@ public AbstractRuntimeUDFContext(String name,
 		this.subtaskIndex = subtaskIndex;
 		this.userCodeClassLoader = userCodeClassLoader;
 		this.executionConfig = executionConfig;
-		this.distributedCache = new DistributedCache(cpTasks);
+		this.distributedCache = Preconditions.checkNotNull(new DistributedCache(cpTasks));
```
Comment (zentol) [inline flink-core/src/main/java/org/apache/flink/api/common/functions/util/AbstractRuntimeUDFContext.java]: Don't you want to check cpTasks for being null?

File: flink-core/src/main/java/org/apache/flink/api/common/functions/util/RuntimeUDFContext.java [hunk 2]
```diff
@@ -37,18 +37,17 @@
...
+		this(name, numParallelSubtasks, subtaskIndex, userCodeClassLoader, executionConfig,
+				new HashMap<String, Future<Path>>(), accumulators);
 	}
-	
+
 	public RuntimeUDFContext(String name, int numParallelSubtasks, int subtaskIndex, ClassLoader userCodeClassLoader,
 							ExecutionConfig executionConfig, Map<String, Future<Path>> cpTasks, Map<String, Accumulator<?,?>> accumulators) {
 		super(name, numParallelSubtasks, subtaskIndex, userCodeClassLoader, executionConfig, accumulators, cpTasks);
 	}
-
```
Comment (zentol) [inline flink-core/src/main/java/org/apache/flink/api/common/functions/util/RuntimeUDFContext.java]: Here we have a few unnecessary formatting changes that just clutter the diff.

File: flink-core/src/main/java/org/apache/flink/api/common/operators/CollectionExecutor.java [hunk 3]
```diff
@@ -501,4 +536,22 @@ public int getSuperstepNumber() {
...
+			this.entry = entry;
+		}
+
+		@Override
+		public Path call() throws IOException{
+			try{
+				LocalFileSystem fs = (LocalFileSystem) entry.getFileSystem();
+				return entry.isAbsolute() ? new Path(entry.toUri().getPath()): new Path(fs.getWorkingDirectory(),entry);
+			} catch (ClassCastException e){
+				throw new RuntimeException("Collection execution must have only local file paths");
```
Comment (zentol) [inline flink-core/src/main/java/org/apache/flink/api/common/operators/CollectionExecutor.java]: dislike this error message, there's is no apparent relation to the distributed cache. 
how about "The DistrbutedCache only supports local files for Collection Environments."

File: flink-core/src/main/java/org/apache/flink/api/common/functions/util/AbstractRuntimeUDFContext.java (hunk 1 above)
Comment (sachingoel0101) [inline flink-core/src/main/java/org/apache/flink/api/common/functions/util/AbstractRuntimeUDFContext.java]: Ah yes. Sorry.

File: flink-core/src/main/java/org/apache/flink/api/common/operators/CollectionExecutor.java (hunk 3 above)
Comment (sachingoel0101) [inline flink-core/src/main/java/org/apache/flink/api/common/operators/CollectionExecutor.java]: Yeah. That would make more sense.

**Comment by sachingoel0101** [comment]
Addressed PR comments. There is one unrelated failure on the GroupReduceITCase. I've filed a JIRA for that.

**Comment by StephanEwen** [comment]
Looks good, in general.

Can you add the test to one of the other iteration test files? This saves cluster startup and shutdown costs, making builds faster. Maybe to the iteration aggregators, or iteration accumulators.

File: flink-core/src/main/java/org/apache/flink/api/common/operators/CollectionExecutor.java [hunk 4]
```diff
@@ -501,4 +536,22 @@ public int getSuperstepNumber() {
 			return (T) previousAggregates.get(name);
 		}
 	}
+
+	private static final class DoingNothing implements Callable<Path>{
```
Comment (StephanEwen) [inline flink-core/src/main/java/org/apache/flink/api/common/operators/CollectionExecutor.java]: It actually does something ;-)

Comment (sachingoel0101) [inline flink-core/src/main/java/org/apache/flink/api/common/operators/CollectionExecutor.java]: Haha. Yes. In an earlier version of the code, it wasn't. :')

**Comment by sachingoel0101** [comment]
I've moved the test to an existing MultipleProgramTestBase`. Should be good to merge now. :)

**Comment by sachingoel0101** [comment]
I'd like to get this merged soon. This removes multiple constructors for Runtime contexts and establishes a clean hierarchy, making any changes to the constructors easier. This will be useful for two Jiras on exposing task configuration and task attempt number to the Runtime context.

**Comment by sachingoel0101** [comment]
I decided to go ahead and implement things which touch the Runtime Context constructors with this PR. This now closes five Jiras, namely 2449, 2458, 2488, 2496 and 2524. Commit messages are descriptive of each Jira.

File: flink-runtime/src/main/scala/org/apache/flink/runtime/taskmanager/TaskManager.scala [hunk 5]
```diff
@@ -897,7 +897,7 @@ class TaskManager(
         config.timeout,
         libCache,
         fileCache,
-        runtimeInfo)
+        new TaskRuntimeInfo(hostname, taskManagerConfig, tdd.getAttemptNumber))
```
Comment (hsaputra) [inline flink-runtime/src/main/scala/org/apache/flink/runtime/taskmanager/TaskManager.scala]: Why is this changed from before?

Comment (sachingoel0101) [inline flink-runtime/src/main/scala/org/apache/flink/runtime/taskmanager/TaskManager.scala]: This is to provide access to Task attempt number from Runtime Context. I should add a description of the other tickets this resolves.
Is this a good idea though? To fix five issues in one PR? Or should I open a separate one and keep this one for just distributed cache?

Comment (zentol) [inline flink-runtime/src/main/scala/org/apache/flink/runtime/taskmanager/TaskManager.scala]: generally we try to keep one PR for one issue, exceptions should only be done for closely related issues.

why did you decide to add these issues into this PR? ( i have a hard time understanding it, since the commits barely touch the same files.

Comment (sachingoel0101) [inline flink-runtime/src/main/scala/org/apache/flink/runtime/taskmanager/TaskManager.scala]: Yes. The addition of distributed cache removes the need for multiple constructors for `RuntimeContext`s. Since providing access to runtime information needed changing the constructors, I deemed it better to work with what would be the only needed constructors after merging this. 
I can revert this commit and open a separate PR for the *other* three issues if necessary.

Comment (zentol) [inline flink-runtime/src/main/scala/org/apache/flink/runtime/taskmanager/TaskManager.scala]: I would prefer if you opened a second PR once this is merged. The issues are not really related to each other; the 2nd commit was simply made based on the 1st commit. We would end up having two separate discussions in 1 PR, which i think is a bad idea.

Comment (sachingoel0101) [inline flink-runtime/src/main/scala/org/apache/flink/runtime/taskmanager/TaskManager.scala]: Ah. Yes. That makes sense. I will revert this and open a separate PR. Apologies.

**Comment by sachingoel0101** [comment]
Reverting back to make this PR only about the distributed cache.

**Comment by StephanEwen** [comment]
We are indeed falling behind on merging pull requests, right now. Many committers are on vacation this month, and for the others, the large amount of pull requests is hard to keep up with, especially next to the work on our own issues.

Hope this will get better in a week or two.

I'll try to get a look at this very soon...

**Comment by StephanEwen** [comment]
In the `CollectionExecutor`, can you skip creating the `ExecutiorService`? You can eagerly resolve the path and then put an already finished future into the map.

**Comment by StephanEwen** [comment]
Aside from the comment above, this looks good. Would merge this, after the comment is addressed.

**Comment by sachingoel0101** [comment]
Addressed comments. @StephanEwen

**Comment by StephanEwen** [comment]
Looks good, merging this!
